File: RootTools2/src/main/java/com/rarnu/tools/neo/loader/ComponentLoader.java
package com.rarnu.tools.neo.loader;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import com.rarnu.tools.neo.base.BaseLoader;
import com.rarnu.tools.neo.data.CompInfo;
import com.rarnu.tools.neo.utils.ComponentUtils;
import com.rarnu.tools.neo.utils.PackageParserUtils;

import java.util.List;

/**
 * Created by rarnu on 9/2/16.
 */
public class ComponentLoader extends BaseLoader<CompInfo> {

    private int type = 0;
    private String pkg = null;

    public ComponentLoader(Context context) {
        super(context);
    }

    @Override
    public List<CompInfo> loadInBackground() {
        List<CompInfo> list = null;
        try {
            ApplicationInfo info = getContext().getPackageManager().getApplicationInfo(pkg, 0);
            PackageParserUtils ppu = new PackageParserUtils();
            Object obj = ppu.parsePackage(info.publicSourceDir, 0);
            switch (type) {
                case 0:
                    list = ComponentUtils.getActivityList(getContext(), obj);
                    break;
                case 1:
                    list = ComponentUtils.getServiceList(getContext(), obj);
                    break;
                case 2:
                    list = ComponentUtils.getReceiverList(getContext(), obj);
                    break;
                case 3:
                    list = ComponentUtils.getProviderList(getContext(), obj);
                    break;
            }
        } catch (Exception e) {
        }
        return list;
    }

    public void startLoading(String pkg, int type) {
        this.type = type;
        this.pkg = pkg;
        startLoading();
    }
}

File: RootTools2/src/main/kotlin/com/rarnu/tools/neo/loader/ComponentLoader.kt
package com.rarnu.tools.neo.loader

import android.content.Context
import android.content.pm.PackageManager
import com.rarnu.android.BaseListLoader
import com.rarnu.tools.neo.utils.CompInfo
import com.rarnu.tools.neo.utils.ComponentUtils

/**
 * Created by rarnu on 9/2/16.
 */
class ComponentLoader(context: Context) : BaseListLoader<CompInfo>(context) {

    private var type = 0
    private var pkg: String? = null

    override fun loadInBackground(): List<CompInfo> {
        val list = mutableListOf<CompInfo>()
        try {
            val info = context.packageManager.getPackageInfo(pkg, PackageManager.GET_ACTIVITIES or PackageManager.GET_SERVICES or PackageManager.GET_RECEIVERS or PackageManager.GET_PROVIDERS)
            when (type) {
                0 -> list.addAll(ComponentUtils.getActivityList(context, info))
                1 -> list.addAll(ComponentUtils.getServiceList(context, info))
                2 -> list.addAll(ComponentUtils.getReceiverList(context, info))
                3 -> list.addAll(ComponentUtils.getProviderList(context, info))
            }
        } catch (e: Exception) {
        }
        return list
    }

    fun startLoading(pkg: String?, type: Int) {
        this.type = type
        this.pkg = pkg
        startLoading()
    }
}
